presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoResultSet.java

```diff
@@ -1840,18 +1845,17 @@ public void interrupt(InterruptedException e)
         }
 
         @Override
-        public boolean hasNext()
+        protected T computeNext()
         {
-            if (!rowQueue.isEmpty()) {
-                return true;
+            try {
+                semaphore.acquire();
             }
-            while (rowQueue.isEmpty() && !future.isDone()) {
-                // making sure rowQueue has some records to process or return false
-                if (parent.isInterrupted()) {
-                    interrupt(new InterruptedException("parent thread interrupted"));
-                }
+            catch (InterruptedException e) {
+                interrupt(e);
             }
-            if (future.isCompletedExceptionally()) {
+            if (rowQueue.isEmpty()) {
+                // If we got here and the queue is empty the future must be done.
+                // Check for ExecutionExceptions.
```

Comment:
Just this:
```
// If we got here and the queue is empty the future must be done. Check status. 
[MoreFutures.]getDone(future);
```

Reply:
Thanks. Can't do the getDone for two reasons:
1. The future might not be done, yet. I.e. the work is done, but it might not be marked done yet, depending on thread scheduling. So we need to wait just in case.
2. The outside code expects to exceptions to be thrown as before

Reply:
I did fix the comment.

Reply:
you're right again. Please update the comment then. It should not be saying "future must be done".

Reply:
Updated the comment again to be more precise.